File: packages/firebase_common/lib/src/firestore_documents/in_review_config.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'in_review_config.dart';

class ReadInReviewConfig {
  const ReadInReviewConfig({
    required this.inReviewConfigId,
    required this.path,
    required this.iOSInReviewVersion,
    required this.enableIOSInReviewMode,
    required this.androidInReviewVersion,
    required this.enableAndroidInReviewMode,
  });

  final String inReviewConfigId;

  final String path;

  final String iOSInReviewVersion;

  final bool enableIOSInReviewMode;

  final String androidInReviewVersion;

  final bool enableAndroidInReviewMode;

  factory ReadInReviewConfig._fromJson(Map<String, dynamic> json) {
    return ReadInReviewConfig(
      inReviewConfigId: json['inReviewConfigId'] as String,
      path: json['path'] as String,
      iOSInReviewVersion: json['iOSInReviewVersion'] as String? ?? '1.0.0',
      enableIOSInReviewMode: json['enableIOSInReviewMode'] as bool? ?? false,
      androidInReviewVersion:
          json['androidInReviewVersion'] as String? ?? '1.0.0',
      enableAndroidInReviewMode:
          json['enableAndroidInReviewMode'] as bool? ?? false,
    );
  }

  factory ReadInReviewConfig.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadInReviewConfig._fromJson(<String, dynamic>{
      ...data,
      'inReviewConfigId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateInReviewConfig {
  const CreateInReviewConfig({
    required this.iOSInReviewVersion,
    required this.enableIOSInReviewMode,
    required this.androidInReviewVersion,
    required this.enableAndroidInReviewMode,
  });

  final String iOSInReviewVersion;
  final bool enableIOSInReviewMode;
  final String androidInReviewVersion;
  final bool enableAndroidInReviewMode;

  Map<String, dynamic> toJson() {
    return {
      'iOSInReviewVersion': iOSInReviewVersion,
      'enableIOSInReviewMode': enableIOSInReviewMode,
      'androidInReviewVersion': androidInReviewVersion,
      'enableAndroidInReviewMode': enableAndroidInReviewMode,
    };
  }
}

class UpdateInReviewConfig {
  const UpdateInReviewConfig({
    this.iOSInReviewVersion,
    this.enableIOSInReviewMode,
    this.androidInReviewVersion,
    this.enableAndroidInReviewMode,
  });

  final String? iOSInReviewVersion;
  final bool? enableIOSInReviewMode;
  final String? androidInReviewVersion;
  final bool? enableAndroidInReviewMode;

  Map<String, dynamic> toJson() {
    return {
      if (iOSInReviewVersion != null) 'iOSInReviewVersion': iOSInReviewVersion,
      if (enableIOSInReviewMode != null)
        'enableIOSInReviewMode': enableIOSInReviewMode,
      if (androidInReviewVersion != null)
        'androidInReviewVersion': androidInReviewVersion,
      if (enableAndroidInReviewMode != null)
        'enableAndroidInReviewMode': enableAndroidInReviewMode,
    };
  }
}

class DeleteInReviewConfig {}

/// Provides a reference to the configurations collection for reading.
final readInReviewConfigCollectionReference = FirebaseFirestore.instance
    .collection('configurations')
    .withConverter<ReadInReviewConfig>(
      fromFirestore: (ds, _) => ReadInReviewConfig.fromDocumentSnapshot(ds),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a inReviewConfig document for reading.
DocumentReference<ReadInReviewConfig> readInReviewConfigDocumentReference({
  required String inReviewConfigId,
}) =>
    readInReviewConfigCollectionReference.doc(inReviewConfigId);

/// Provides a reference to the configurations collection for creating.
final createInReviewConfigCollectionReference = FirebaseFirestore.instance
    .collection('configurations')
    .withConverter<CreateInReviewConfig>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a inReviewConfig document for creating.
DocumentReference<CreateInReviewConfig> createInReviewConfigDocumentReference({
  required String inReviewConfigId,
}) =>
    createInReviewConfigCollectionReference.doc(inReviewConfigId);

/// Provides a reference to the configurations collection for updating.
final updateInReviewConfigCollectionReference = FirebaseFirestore.instance
    .collection('configurations')
    .withConverter<UpdateInReviewConfig>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (obj, _) => obj.toJson(),
    );

/// Provides a reference to a inReviewConfig document for updating.
DocumentReference<UpdateInReviewConfig> updateInReviewConfigDocumentReference({
  required String inReviewConfigId,
}) =>
    updateInReviewConfigCollectionReference.doc(inReviewConfigId);

/// Provides a reference to the configurations collection for deleting.
final deleteInReviewConfigCollectionReference = FirebaseFirestore.instance
    .collection('configurations')
    .withConverter<DeleteInReviewConfig>(
      fromFirestore: (_, __) => throw UnimplementedError(),
      toFirestore: (_, __) => throw UnimplementedError(),
    );

/// Provides a reference to a inReviewConfig document for deleting.
DocumentReference<DeleteInReviewConfig> deleteInReviewConfigDocumentReference({
  required String inReviewConfigId,
}) =>
    deleteInReviewConfigCollectionReference.doc(inReviewConfigId);

/// Manages queries against the configurations collection.
class InReviewConfigQuery {
  /// Fetches [ReadInReviewConfig] documents.
  Future<List<ReadInReviewConfig>> fetchDocuments({
    GetOptions? options,
    Query<ReadInReviewConfig>? Function(Query<ReadInReviewConfig> query)?
        queryBuilder,
    int Function(ReadInReviewConfig lhs, ReadInReviewConfig rhs)? compare,
  }) async {
    Query<ReadInReviewConfig> query = readInReviewConfigCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [InReviewConfig] documents.
  Stream<List<ReadInReviewConfig>> subscribeDocuments({
    Query<ReadInReviewConfig>? Function(Query<ReadInReviewConfig> query)?
        queryBuilder,
    int Function(ReadInReviewConfig lhs, ReadInReviewConfig rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadInReviewConfig> query = readInReviewConfigCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadInReviewConfig] document.
  Future<ReadInReviewConfig?> fetchDocument({
    required String inReviewConfigId,
    GetOptions? options,
  }) async {
    final ds = await readInReviewConfigDocumentReference(
      inReviewConfigId: inReviewConfigId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [InReviewConfig] document.
  Stream<ReadInReviewConfig?> subscribeDocument({
    required String inReviewConfigId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readInReviewConfigDocumentReference(
      inReviewConfigId: inReviewConfigId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [InReviewConfig] document.
  Future<DocumentReference<CreateInReviewConfig>> add({
    required CreateInReviewConfig createInReviewConfig,
  }) =>
      createInReviewConfigCollectionReference.add(createInReviewConfig);

  /// Sets a [InReviewConfig] document.
  Future<void> set({
    required String inReviewConfigId,
    required CreateInReviewConfig createInReviewConfig,
    SetOptions? options,
  }) =>
      createInReviewConfigDocumentReference(
        inReviewConfigId: inReviewConfigId,
      ).set(createInReviewConfig, options);

  /// Updates a specific [InReviewConfig] document.
  Future<void> update({
    required String inReviewConfigId,
    required UpdateInReviewConfig updateInReviewConfig,
  }) =>
      updateInReviewConfigDocumentReference(
        inReviewConfigId: inReviewConfigId,
      ).update(updateInReviewConfig.toJson());

  /// Deletes a specific [InReviewConfig] document.
  Future<void> delete({
    required String inReviewConfigId,
  }) =>
      deleteInReviewConfigDocumentReference(
        inReviewConfigId: inReviewConfigId,
      ).delete();
}

File: functions/src/firestore-documents/inReviewConfig.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadInReviewConfig {
    constructor({
        inReviewConfigId,
        path,
        iOSInReviewVersion,
        enableIOSInReviewMode,
        androidInReviewVersion,
        enableAndroidInReviewMode
    }: {
        inReviewConfigId: string
        path: string
        iOSInReviewVersion: string
        enableIOSInReviewMode: boolean
        androidInReviewVersion: string
        enableAndroidInReviewMode: boolean
    }) {
        this.inReviewConfigId = inReviewConfigId
        this.path = path
        this.iOSInReviewVersion = iOSInReviewVersion
        this.enableIOSInReviewMode = enableIOSInReviewMode
        this.androidInReviewVersion = androidInReviewVersion
        this.enableAndroidInReviewMode = enableAndroidInReviewMode
    }

    readonly inReviewConfigId: string

    readonly path: string

    readonly iOSInReviewVersion: string

    readonly enableIOSInReviewMode: boolean

    readonly androidInReviewVersion: string

    readonly enableAndroidInReviewMode: boolean

    private static fromJson(json: Record<string, unknown>): ReadInReviewConfig {
        return new ReadInReviewConfig({
            inReviewConfigId: json[`inReviewConfigId`] as string,
            path: json[`path`] as string,
            iOSInReviewVersion:
                (json[`iOSInReviewVersion`] as string | undefined) ?? `1.0.0`,
            enableIOSInReviewMode:
                (json[`enableIOSInReviewMode`] as boolean | undefined) ?? false,
            androidInReviewVersion:
                (json[`androidInReviewVersion`] as string | undefined) ??
                `1.0.0`,
            enableAndroidInReviewMode:
                (json[`enableAndroidInReviewMode`] as boolean | undefined) ??
                false
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadInReviewConfig {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadInReviewConfig.fromJson({
            ...cleanedData,
            inReviewConfigId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateInReviewConfig {
    constructor({
        iOSInReviewVersion,
        enableIOSInReviewMode,
        androidInReviewVersion,
        enableAndroidInReviewMode
    }: {
        iOSInReviewVersion: string
        enableIOSInReviewMode: boolean
        androidInReviewVersion: string
        enableAndroidInReviewMode: boolean
    }) {
        this.iOSInReviewVersion = iOSInReviewVersion
        this.enableIOSInReviewMode = enableIOSInReviewMode
        this.androidInReviewVersion = androidInReviewVersion
        this.enableAndroidInReviewMode = enableAndroidInReviewMode
    }

    readonly iOSInReviewVersion: string

    readonly enableIOSInReviewMode: boolean

    readonly androidInReviewVersion: string

    readonly enableAndroidInReviewMode: boolean

    toJson(): Record<string, unknown> {
        return {
            iOSInReviewVersion: this.iOSInReviewVersion,
            enableIOSInReviewMode: this.enableIOSInReviewMode,
            androidInReviewVersion: this.androidInReviewVersion,
            enableAndroidInReviewMode: this.enableAndroidInReviewMode
        }
    }
}

export class UpdateInReviewConfig {
    constructor({
        iOSInReviewVersion,
        enableIOSInReviewMode,
        androidInReviewVersion,
        enableAndroidInReviewMode
    }: {
        iOSInReviewVersion?: string
        enableIOSInReviewMode?: boolean
        androidInReviewVersion?: string
        enableAndroidInReviewMode?: boolean
    }) {
        this.iOSInReviewVersion = iOSInReviewVersion
        this.enableIOSInReviewMode = enableIOSInReviewMode
        this.androidInReviewVersion = androidInReviewVersion
        this.enableAndroidInReviewMode = enableAndroidInReviewMode
    }

    readonly iOSInReviewVersion?: string

    readonly enableIOSInReviewMode?: boolean

    readonly androidInReviewVersion?: string

    readonly enableAndroidInReviewMode?: boolean

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.iOSInReviewVersion != undefined) {
            json[`iOSInReviewVersion`] = this.iOSInReviewVersion
        }
        if (this.enableIOSInReviewMode != undefined) {
            json[`enableIOSInReviewMode`] = this.enableIOSInReviewMode
        }
        if (this.androidInReviewVersion != undefined) {
            json[`androidInReviewVersion`] = this.androidInReviewVersion
        }
        if (this.enableAndroidInReviewMode != undefined) {
            json[`enableAndroidInReviewMode`] = this.enableAndroidInReviewMode
        }

        return json
    }
}

export class DeleteInReviewConfig {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the configurations collection for reading.
 */
export const readInReviewConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<ReadInReviewConfig>({
        fromFirestore: (ds: DocumentSnapshot): ReadInReviewConfig => {
            return ReadInReviewConfig.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(
                `toFirestore is not implemented for ReadInReviewConfig`
            )
        }
    })

/**
 * Provides a reference to a inReviewConfig document for reading.
 * @param inReviewConfigId - The ID of the inReviewConfig document to read.
 */
export const readInReviewConfigDocumentReference = ({
    inReviewConfigId
}: {
    inReviewConfigId: string
}): DocumentReference<ReadInReviewConfig> =>
    readInReviewConfigCollectionReference.doc(inReviewConfigId)

/**
 * Provides a reference to the configurations collection for creating.
 */
export const createInReviewConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<CreateInReviewConfig>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for CreateInReviewConfig`
            )
        },
        toFirestore: (obj: CreateInReviewConfig): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a inReviewConfig document for creating.
 * @param inReviewConfigId - The ID of the inReviewConfig document to read.
 */
export const createInReviewConfigDocumentReference = ({
    inReviewConfigId
}: {
    inReviewConfigId: string
}): DocumentReference<CreateInReviewConfig> =>
    createInReviewConfigCollectionReference.doc(inReviewConfigId)

/**
 * Provides a reference to the configurations collection for updating.
 */
export const updateInReviewConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<UpdateInReviewConfig>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for UpdateInReviewConfig`
            )
        },
        toFirestore: (obj: UpdateInReviewConfig): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a inReviewConfig document for updating.
 * @param inReviewConfigId - The ID of the inReviewConfig document to read.
 */
export const updateInReviewConfigDocumentReference = ({
    inReviewConfigId
}: {
    inReviewConfigId: string
}): DocumentReference<UpdateInReviewConfig> =>
    updateInReviewConfigCollectionReference.doc(inReviewConfigId)

/**
 * Provides a reference to the configurations collection for deleting.
 */
export const deleteInReviewConfigCollectionReference = db
    .collection(`configurations`)
    .withConverter<DeleteInReviewConfig>({
        fromFirestore: () => {
            throw new Error(
                `fromFirestore is not implemented for DeleteInReviewConfig`
            )
        },
        toFirestore: (): DocumentData => {
            throw new Error(
                `toFirestore is not implemented for DeleteInReviewConfig`
            )
        }
    })

/**
 * Provides a reference to a inReviewConfig document for deleting.
 * @param inReviewConfigId - The ID of the inReviewConfig document to read.
 */
export const deleteInReviewConfigDocumentReference = ({
    inReviewConfigId
}: {
    inReviewConfigId: string
}): DocumentReference<DeleteInReviewConfig> =>
    deleteInReviewConfigCollectionReference.doc(inReviewConfigId)

/**
 * Manages queries against the configurations collection.
 */
export class InReviewConfigQuery {
    /**
     * Fetches inReviewConfig documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (
            query: Query<ReadInReviewConfig>
        ) => Query<ReadInReviewConfig>
        compare?: (lhs: ReadInReviewConfig, rhs: ReadInReviewConfig) => number
    }): Promise<ReadInReviewConfig[]> {
        let query: Query<ReadInReviewConfig> =
            readInReviewConfigCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadInReviewConfig> = await query.get()
        let result = qs.docs.map(
            (qds: QueryDocumentSnapshot<ReadInReviewConfig>) => qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific inReviewConfig document.
     * @param inReviewConfigId - The ID of the inReviewConfig document to fetch.
     */
    async fetchDocument({
        inReviewConfigId
    }: {
        inReviewConfigId: string
    }): Promise<ReadInReviewConfig | undefined> {
        const ds = await readInReviewConfigDocumentReference({
            inReviewConfigId
        }).get()
        return ds.data()
    }

    /**
     * Adds a inReviewConfig document.
     * @param createInReviewConfig - The inReviewConfig details to add.
     */
    async add({
        createInReviewConfig
    }: {
        createInReviewConfig: CreateInReviewConfig
    }): Promise<DocumentReference<CreateInReviewConfig>> {
        return createInReviewConfigCollectionReference.add(createInReviewConfig)
    }

    /**
     * Sets a inReviewConfig document.
     * @param inReviewConfigId - The ID of the inReviewConfig document to set.
     * @param createInReviewConfig - The inReviewConfig details to set.
     * @param options - Options for the set operation.
     */
    async set({
        inReviewConfigId,
        createInReviewConfig,
        options
    }: {
        inReviewConfigId: string
        createInReviewConfig: CreateInReviewConfig
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createInReviewConfigDocumentReference({
                inReviewConfigId
            }).set(createInReviewConfig)
        } else {
            return createInReviewConfigDocumentReference({
                inReviewConfigId
            }).set(createInReviewConfig, options)
        }
    }

    /**
     * Updates a specific inReviewConfig document.
     * @param inReviewConfigId - The ID of the inReviewConfig document to update.
     * @param updateInReviewConfig - The details for updating the inReviewConfig.
     */
    async update({
        inReviewConfigId,
        updateInReviewConfig
    }: {
        inReviewConfigId: string
        updateInReviewConfig: UpdateInReviewConfig
    }): Promise<WriteResult> {
        return updateInReviewConfigDocumentReference({
            inReviewConfigId
        }).update(updateInReviewConfig.toJson())
    }

    /**
     * Deletes a specific inReviewConfig document.
     * @param inReviewConfigId - The ID of the inReviewConfig document to delete.
     */
    async delete({
        inReviewConfigId
    }: {
        inReviewConfigId: string
    }): Promise<WriteResult> {
        return deleteInReviewConfigDocumentReference({
            inReviewConfigId
        }).delete()
    }
}
